src/node.rs

```diff
@@ -114,16 +126,263 @@ impl InMemoryNodeInner {
+        };
+
+        let tx: Transaction = l2_tx.clone().into();
+        let l1_gas_price =
+            adjust_l1_gas_price_for_tx(l1_gas_price, L2_GAS_PRICE, tx.gas_per_pubdata_byte_limit());
+
+        // Set gas_limit for transaction
+        let gas_limit_with_overhead = tx_gas_limit
+            + derive_gas_estimation_overhead(
+                tx_gas_limit,
+                gas_per_pubdata_byte as u32,
+                tx.encoding_len(),
+            );
+        l2_tx.common_data.fee.gas_limit = gas_limit_with_overhead.into();
+
+        let fork_storage_copy = self.fork_storage.clone();
+        let mut storage_view = StorageView::new(&fork_storage_copy);
+
+        // The nonce needs to be updated
+        let nonce = l2_tx.nonce();
+        let nonce_key = get_nonce_key(&l2_tx.initiator_account());
+        let full_nonce = storage_view.read_value(&nonce_key);
+        let (_, deployment_nonce) = decompose_full_nonce(h256_to_u256(full_nonce));
+        let enforced_full_nonce = nonces_to_full_nonce(U256::from(nonce.0), deployment_nonce);
+        storage_view.set_value(nonce_key, u256_to_h256(enforced_full_nonce));
+
+        // We need to explicitly put enough balance into the account of the users
+        let payer = l2_tx.payer();
+        let balance_key = storage_key_for_eth_balance(&payer);
+        let mut current_balance = h256_to_u256(storage_view.read_value(&balance_key));
+        let added_balance = l2_tx.common_data.fee.gas_limit * l2_tx.common_data.fee.max_fee_per_gas;
+        current_balance += added_balance;
+        storage_view.set_value(balance_key, u256_to_h256(current_balance));
+
+        let mut oracle_tools = OracleTools::new(&mut storage_view, HistoryDisabled);
+
+        let mut block_context = self.create_block_context();
+        block_context.l1_gas_price = l1_gas_price;
+        let derived_block_context = DerivedBlockContext {
+            context: self.create_block_context(),
+            base_fee,
+        };
+
+        // Use the fee_estimate bootloader code, as it sets ENSURE_RETURNED_MAGIC to 0 and BOOTLOADER_TYPE to 'playground_block'
+        let bootloader_code = &self.fee_estimate_contracts;
+        let block_properties = InMemoryNodeInner::create_block_properties(bootloader_code);
+
+        // init vm
+        let mut vm = init_vm_inner(
+            &mut oracle_tools,
+            BlockContextMode::OverrideCurrent(derived_block_context),
+            &block_properties,
+            BLOCK_GAS_LIMIT,
+            bootloader_code,
+            execution_mode,
+        );
+
+        let tx: Transaction = l2_tx.into();
+
+        push_transaction_to_bootloader_memory(&mut vm, &tx, execution_mode, None);
+
+        let vm_block_result = vm.execute_till_block_end(BootloaderJobType::TransactionExecution);
+
+        match vm_block_result.full_result.revert_reason {
+            None => Ok(vm_block_result),
+            Some(revert) => Err(revert.revert_reason),
+        }
```

Comment:
We seem to do this same order of operations in a number of different places, consider making this a function that can be reused in `run_l2_call`, `run_l2_tx_inner`, and `estimate_gas_step`

Reply:
I'd say let's do that in a follow up PR if possible. These implementations are slightly different enough that to consolidate would touch/add a lot of code.